source/extensions/filters/udp/dns_filter/dns_filter.cc

```diff
@@ -76,21 +109,149 @@ void DnsFilter::onData(Network::UdpRecvData& client_request) {
     return;
   }
 
-  // TODO(abaptiste): Resolve the requested name
+  // Resolve the requested name
+  const auto response = getResponseForQuery(query_context);
 
-  // Send an answer to the client
+  // We were not able to satisfy the request locally. Return an empty response to the client
+  if (response == DnsLookupResponseCode::Failure) {
+    sendDnsResponse(std::move(query_context));
+    return;
+  }
+
+  // TODO(abaptiste): external resolution
+
+  // We have an answer. Send it to the client
   sendDnsResponse(std::move(query_context));
 }
 
 void DnsFilter::sendDnsResponse(DnsQueryContextPtr query_context) {
   Buffer::OwnedImpl response;
-  // TODO(abaptiste): serialize and return a response to the client
 
+  // Serializes the generated response to the parsed query from the client. If there is a
+  // parsing error or the incoming query is invalid, we will still generate a valid DNS response
+  message_parser_.buildResponseBuffer(query_context, response);
   Network::UdpSendData response_data{query_context->local_->ip(), *(query_context->peer_),
                                      response};
   listener_.send(response_data);
 }
 
+DnsLookupResponseCode DnsFilter::getResponseForQuery(DnsQueryContextPtr& context) {
+  /* It appears to be a rare case where we would have more than one query in a single request.
+   * It is allowed by the protocol but not widely supported:
+   *
+   * See: https://www.ietf.org/rfc/rfc1035.txt
+   *
+   * The question section is used to carry the "question" in most queries,
+   * i.e., the parameters that define what is being asked. The section
+   * contains QDCOUNT (usually 1) entries.
+   */
+  for (const auto& query : context->queries_) {
+    // Try to resolve the query locally. If forwarding the query externally is disabled we will
```

Comment:
can a list of query names have a mix of resolving them externally and internally? I think supporting that will be harder. we must either have only external queries or just internal IMO

Reply:
I agree.  

The way the filter works now is that the first query to "resolve" in the internal/external path will "win".  If the filter returns "Success" from parsing the query and proceeds with the response generation, then the client will have one of the queries answered (this will likely be the internal query since there will be some latency do to the external resolution).

Supporting that case would be difficult.  C-ares will return an error if it finds more than one question in a query, so we could go that route and enforce that as well.

https://github.com/c-ares/c-ares/blob/7ebedab25dab50b2f008fbef8601c223096bb780/ares_parse_ns_reply.c#L66

This is just one ares instance of the query count check.

That said, I do have a test where I have multiple internal queries where both questions are answered in a single response.   This is still rare in the wild.

Reply:
I thought the change is minor - Isn't it? The oneof needs a string for cluster_name, which is used in this line referenced by the comment right?